test/e2e-test/rule-test/util/test-rule-functions.js

```diff
@@ -0,0 +1,145 @@
+    else
+        return response(true, '', reqJson, resJson);
+};
+
+var noDefaultGuid = function(reqJson, resJson){
+    var ids = fhir.evaluate(resJson, 'Bundle.entry.resource.id');
+    var defaultGuidCount = testUtils.countOccurences(ids, testUtils.defaultGuid);
+    if(defaultGuidCount >= 1)
+        return response(false, 'The bundle contains ' + defaultGuidCount + ' default Guid(s) ' + testUtils.defaultGuid);
+    else
+        return response(true, '', reqJson, resJson);
+};
+
+var noSameGuid = function(reqJson, resJson){
+    var resources = fhir.evaluate(resJson, 'Bundle.entry.resource');
+    var ids = [];
+    for(var index in resources){
+        ids.push(resources[index].resourceType + '/' + resources[index].id);
+    }
+    var duplicates = testUtils.findDuplicates(ids);
+    if(duplicates.length !== 0)
+        return response(false, 'The bundle contains some duplicate Guids: ' + duplicates.toString());
+    else
+        return response(true, '', reqJson, resJson);
+};
+
+const __revealObjectValues = (target, object, level) => {
+    if (level >= MAX_REVEAL_DEPTH) {
+        throw new Error('Reveal depth exceeds limit.');
+    }
+    if (_.isObject(object)) {
+        const keys = Object.keys(object).filter(key => !NON_COMPARE_PROPERTY.has(key));
+        return keys.every(key => __revealObjectValues(target, object[key], level + 1));
+    }
+    else {
+        const value = object.toString();
+        // specially treate datetime data type
+        const dateTimeRegex = /^[1-9]\d{3}-(0[1-9]|1[0-2])-(0[1-9]|[1-2][0-9]|3[0-1])(\s+)?T?((20|21|22|23|[0-1]\d):[0-5]\d(:[0-5]\d)?)?.*$/;
+        if (dateTimeRegex.test(value)) {
+            return true;
+        }
+        return target.includes(value);
+    }
+};
+
+/**
+ * Every required property value appeared in conversion result should also appear in origin data.
+ * Some special data type should be treated separately (for example, "DataTime", since it has been processed by helper function).
+ * TODO: But there are many property values are extended from mapping table's explicit attribute, so there need a better way.
+ */
+const backwardValueReveal = (reqJson, resJson) => {
+    try {
+        const flag = __revealObjectValues(JSON.stringify(reqJson), resJson, 0);
+        const message = flag ? '' : 'Some properties can\'t be found in the origin data.';
+        return response(flag, message, reqJson, resJson);
+    }
+    catch (error) {
+        return response(false, error.toString(), reqJson, resJson);
+    }
+};
+
+/**
+ * Use officially recommended validator to validate resources.
+ */
+const officialValidator = (reqJson, resJson) => {
+    const javaExistCommand = 'java -XshowSettings:properties -version';
+    const validatorPath = path.join(__dirname, '../lib/validator_cli.jar');
+    const resourceFolder = path.join(__dirname, '../test-samples/tmp');
+    const resourcePath = path.join(resourceFolder, `${uuidv4().replace(/-/g, '')}.json`);
```

Comment:
What does uuid.v4().replace() used for?

Reply:
It's used to generate a 32-bit unique hex string, which is a part of temporary file path.

Reply:
Do I understand it correctly that this file is used to store `resJson` temporarily?

Reply:
Correct, and it will be removed immediately after validation.